src/k8s_upgrade.py

```diff
@@ -0,0 +1,485 @@
+
+
+class _PostUpgradeCheckMongoDB(EventBase):
+    """Run post upgrade check on MongoDB to verify that the cluster is healhty."""
+
+    def __init__(self, handle):
+        super().__init__(handle)
+
+
+class MongoDBUpgrade(GenericMongoDBUpgrade):
+    """Handlers for upgrade events."""
+
+    post_app_upgrade_event = EventSource(_PostUpgradeCheckMongoDB)
+    post_cluster_upgrade_event = EventSource(_PostUpgradeCheckMongoDB)
+
+    def __init__(self, charm: CharmBase):
+        self.charm = charm
+        super().__init__(charm, PEER_RELATION_ENDPOINT_NAME)
+
+    @override
+    def _observe_events(self, charm: CharmBase) -> None:
+        self.framework.observe(
+            charm.on[PRECHECK_ACTION_NAME].action, self._on_pre_upgrade_check_action
+        )
+        self.framework.observe(
+            charm.on[PEER_RELATION_ENDPOINT_NAME].relation_created,
+            self._on_upgrade_peer_relation_created,
+        )
+        self.framework.observe(charm.on.upgrade_charm, self._on_upgrade)
+        self.framework.observe(
+            charm.on[PEER_RELATION_ENDPOINT_NAME].relation_changed, self._reconcile_upgrade
+        )
+        self.framework.observe(charm.on[RESUME_ACTION_NAME].action, self._on_resume_upgrade_action)
+        self.framework.observe(self.post_app_upgrade_event, self.run_post_app_upgrade_task)
+        self.framework.observe(self.post_cluster_upgrade_event, self.run_post_cluster_upgrade_task)
+
+    def _on_upgrade(self):
+        if self.charm.unit.is_leader():
+            self.charm.version_checker.set_version_across_all_relations()
+        self._upgrade.save_revision()
+
+    def _reconcile_upgrade(self, event: EventBase) -> None:
+        """Handle upgrade events."""
+        if not self._upgrade:
+            logger.debug("Peer relation not available")
+            return
+        if not self._upgrade.versions_set:
+            logger.debug("Peer relation not ready")
+            return
+        if self.charm.unit.is_leader() and not self._upgrade.in_progress:
+            # Run before checking `self._upgrade.is_compatible` in case incompatible upgrade was
+            # forced & completed on all units.
+            self._upgrade.set_versions_in_app_databag()
+
+        if self._upgrade.unit_state is UnitState.RESTARTING:  # Kubernetes only
+            if not self._upgrade.is_compatible:
+                logger.info(
+                    "Upgrade incompatible. If you accept potential *data loss* and *downtime*, you can continue with `resume-upgrade force=true`"
+                )
+                self.charm.status.set_and_share_status(Config.Status.UNHEALTHY_UPGRADE)
+                return
+        if not self._upgrade.unit_state:
+            self._upgrade.unit_state = UnitState.HEALTHY
+        if self.charm.unit.is_leader():
+            self._upgrade.reconcile_partition()
+        if self._upgrade.unit_state is UnitState.RESTARTING:
+            self.post_app_upgrade_event.emit()
```

Comment:
The post app upgrade event requires the underlying mongodb application to be running and performs some sanity checks: check that we can still read/write, that we can communicate with the cluster, and processes the statuses of all required units to validate that the application works as expected.

Reply:
`upgrade-charm` event runs before the unit has fully restarted right? I need to check but I'm unsure about the order here.
If it runs after that, then yeah I can bring this code there.

Reply:
It seems the order of hooks in case of upgrade is `upgrade-charm -> start -> pebble-ready`.
I'm wondering what I should do then since these checks needs mongodb to be running which I expect to be set by the `start` event or the `pebble_ready` event. I could start services in the upgrade-charm event (which would look very much like the pebble ready event) and then run the checks?